app/models/loan.rb

```diff
@@ -189,4 +189,49 @@ def ensure_currency
   def amount_formatted
     ensure_currency.format_amount(amount)
   end
+
+  def signed_contract?
+    media.where(kind_value: 'contract').count > 0
+  end
+
+  def active?
+    status_value == 'active'
+  end
+
+  def incomplete_steps?
+    timeline_entries.merge(ProjectStep.past_due).count > 0
+  end
+
+  def sporadic_loan_updates?
+    timeline_entries.merge(ProjectStep.recent).count < [days_old, 30].min
+  end
+
+  def progress_pct
+    return 0 unless criteria
+    criteria.progress_pct
+  end
+
+  def health_warnings
+    warnings = []
+    warnings << :active_without_signed_contract if active? && !signed_contract?
+    warnings << :active_without_recent_logs if active? && most_recent_log_date < 30.days.ago
+    warnings << :past_due_steps if incomplete_steps?
+    warnings << :incomplete_loan_questions if progress_pct < 80
+    warnings << :sporadic_loan_updates if sporadic_loan_updates?
+    warnings
+  end
+
+  def healthy?
+    health_warnings.count < 1
+  end
+
+  def days_old
+    (end_date.beginning_of_day - signing_date.beginning_of_day) / (24 * 60 * 60)
+  end
+
+  private
+
+  def most_recent_log_date
+    project_logs.maximum(:date) || Time.at(0)
+  end
```

Comment:
I think the best way around this is probably to cache all the necessary values and update them whenever the appropriate records are updated.

So...

* When media are updated, cache whether there is a contract
* When a step is saved, cache whether there are any past due or recently updated steps
* When a ResponseSet is updated, calculate the progress_pct and cache that (and also perhaps refactor the Loan details page to use this value instead of calculating directly)
* When a log is created, update the max log date

If all these things are stored as columns in the projects table, then calculating the health checks can be done entirely in memory once the Loans are loaded on the dashboard.

If you agree with these changes, I would like to be involved in designing how they will be implemented because I am worried about a bunch of callback couplings and things getting brittle and scary... I have some ideas on how to avoid that.